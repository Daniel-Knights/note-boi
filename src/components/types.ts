import { Ref } from 'vue';

export type DropMenuItemData = {
  label: string;
  clickHandler?: () => void;
  testId?: string;
  disabled?: boolean;
  selected?: boolean | Ref<boolean>;
  danger?: boolean;
  subMenu?: DropMenuItemData[];
};
