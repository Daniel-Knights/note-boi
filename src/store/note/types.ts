import { DeletedNote } from '../../api';

export type UnsyncedEventDetail =
  | {
      note: string;
      kind: 'edited';
    }
  | {
      note: DeletedNote;
      kind: 'deleted';
    };
