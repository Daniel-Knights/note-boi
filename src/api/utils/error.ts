import { AppError, ERROR_CODE, ErrorConfig } from '../../classes';

/** Catches hanging requests (e.g. due to server error). */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwFetchError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    message: 'Request failed',
    originalError,
  });
}

/** Catches note encryption errors. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwEncryptorError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'code' | 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    code: ERROR_CODE.ENCRYPTOR,
    message: 'Note encryption/decryption failed',
    originalError,
  });
}
