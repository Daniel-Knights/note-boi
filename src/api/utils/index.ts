export * from './decorators';
export * from './error';
export * from './http';
