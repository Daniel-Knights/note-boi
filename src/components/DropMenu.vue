<template>
  <ul v-if="show" class="drop-menu">
    <DropMenuItem :items="items" />
  </ul>
</template>

<script lang="ts" setup>
import { onMounted, ref } from 'vue';

import { DropMenuItemData } from './types';

import DropMenuItem from './DropMenuItem.vue';

defineProps<{ items: DropMenuItemData[] }>();

const show = ref(true);

function hide() {
  show.value = false;
}

onMounted(() => {
  document.addEventListener('click', hide, { once: true });
});
</script>

<style lang="scss">
$list-bg-colour: var(--colour__interactive);

.drop-menu {
  &,
  ul {
    position: absolute;
    width: 133px;
    color: var(--colour__white);
    background-color: $list-bg-colour;
    box-shadow: 0 0 5px rgba(0, 0, 0, 0.5);
    z-index: 10;
  }

  li {
    cursor: pointer;
    padding: 0.5em 1em;
    white-space: nowrap;
    border: v.$drop-menu-padding solid $list-bg-colour;

    &:hover {
      background-color: var(--colour__tertiary);
    }
  }
}
</style>
