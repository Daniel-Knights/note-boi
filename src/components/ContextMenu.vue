<template>
  <ul v-if="show" class="context-menu" :style="{ top: top + 'px', left: left + 'px' }">
    <li @click="newNote">New Note</li>
    <li
      :class="{ 'context-menu__item--disabled': comp?.hasOneEmptyNote }"
      @click="handleDeleteNote"
    >
      Delete Note
    </li>
    <li class="context-menu__has-sub-menu">
      Theme
      <ul>
        <li
          v-for="theme in colourThemes"
          :key="theme"
          :class="{ 'context-menu__item--selected': theme === selectedTheme }"
          @click="setTheme(theme)"
        >
          {{ theme }}
        </li>
      </ul>
    </li>
    <li class="context-menu__has-sub-menu">
      Auto-sync
      <ul>
        <li
          :class="{ 'context-menu__item--selected': syncState.autoSyncEnabled }"
          @click="setAutoSync(true)"
        >
          On
        </li>
        <li
          :class="{ 'context-menu__item--selected': !syncState.autoSyncEnabled }"
          @click="setAutoSync(false)"
        >
          Off
        </li>
      </ul>
    </li>
  </ul>
</template>

<script lang="ts" setup>
import { computed, ref, watch } from 'vue';

import {
  state as noteState,
  findNote,
  newNote,
  deleteNote,
  deleteAllNotes,
} from '../store/note';
import { state as syncState, setAutoSync } from '../store/sync';
import { isEmptyNote } from '../utils';

const clickedNoteId = ref<string | undefined>(undefined);
const show = ref(false);
const top = ref(0);
const left = ref(0);

const props = defineProps({
  ev: {
    type: MouseEvent || undefined,
    default: undefined,
  },
});

const comp = computed(() => {
  if (!clickedNoteId.value) return;

  const foundNote = findNote(clickedNoteId.value);
  if (!foundNote) return;

  return {
    hasOneEmptyNote: isEmptyNote(foundNote),
  };
});

type Theme = 'Light' | 'Dark' | 'System';

const colourThemes: Theme[] = ['Light', 'Dark', 'System'];
const selectedTheme = ref(localStorage.getItem('theme') || 'System');

document.body.classList.add(`theme--${selectedTheme.value.toLowerCase()}`);

function setTheme(theme: Theme) {
  document.body.classList.remove(`theme--${selectedTheme.value.toLowerCase()}`);
  document.body.classList.add(`theme--${theme.toLowerCase()}`);

  localStorage.setItem('theme', theme);

  selectedTheme.value = theme;
}

function handleDeleteNote() {
  // If multiple notes are selected, delete all of them
  if (noteState.extraSelectedNotes.length > 0) {
    deleteAllNotes();
  } else if (clickedNoteId.value) {
    // Delete clicked note and select next note if currently selected
    deleteNote(clickedNoteId.value, clickedNoteId.value === noteState.selectedNote.id);
  }
}

function hide() {
  show.value = false;
  document.removeEventListener('click', hide);
}

watch(props, () => {
  if (!props.ev) return;

  const target = props.ev.target as HTMLElement | null;
  const closestNote = target?.closest<HTMLElement>('[data-note-id]');

  show.value = true;
  top.value = props.ev.clientY;
  left.value = props.ev.clientX;
  clickedNoteId.value = closestNote?.dataset.noteId;

  document.addEventListener('click', hide);
});
</script>

<style lang="scss" scoped>
$list-padding: 3px;
$list-bg-colour: var(--colour__interactive);

ul {
  position: absolute;
  width: 133px;
  color: var(--colour__white);
  background-color: $list-bg-colour;
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.5);
  z-index: 10;
}

li {
  cursor: pointer;
  padding: 0.5em 1em;
  white-space: nowrap;
  border: $list-padding solid $list-bg-colour;

  &:hover {
    background-color: var(--colour__tertiary);
  }
}

.context-menu__item--disabled {
  pointer-events: none;
  color: var(--colour__tertiary);
}

.context-menu__has-sub-menu {
  position: relative;

  > ul {
    display: none;
    top: -$list-padding;
    left: calc(100% + $list-padding);
  }

  &:hover > ul {
    display: block;
  }
}

.context-menu__item--selected {
  background-color: var(--colour__tertiary);
}
</style>
