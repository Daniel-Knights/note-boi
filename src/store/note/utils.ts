import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
