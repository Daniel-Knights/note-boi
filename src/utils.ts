import { event, invoke } from '@tauri-apps/api';
import { EventCallback, UnlistenFn } from '@tauri-apps/api/event';

import { TauriCommand } from './constant';
import type { Note } from './store/note';

/** `process.env.NODE_ENV === 'development'`. */
export function isDev(): boolean {
  return process.env.NODE_ENV === 'development';
}

/** Formats Unix time to date-time. */
export function unixToDateTime(unixTime: number): string {
  return Intl.DateTimeFormat([], {
    dateStyle: 'medium',
    timeStyle: isDev() ? 'long' : 'short',
  }).format(unixTime);
}

/** Gets key value from `localStorage` and parses it as JSON. */
export function localStorageParse(key: string): ReturnType<typeof JSON.parse> {
  return JSON.parse(localStorage.getItem(key) || '{}');
}

/** Returns `true` if string consists of only whitespace characters or is empty. */
export function isWhitespaceOnly(text?: string): boolean {
  return text?.trim() === '';
}

/** Returns true if note content is empty or note is `undefined`. */
export function isEmptyNote(note?: Note): boolean {
  if (!note) return true;

  return isWhitespaceOnly(note.content.title) && isWhitespaceOnly(note.content.body);
}

/** Calls {@link event.emit}, with stronger typing for `id`. */
export function tauriEmit<T>(id: 'login' | 'logout', payload?: T): void {
  event.emit(id, payload);
}

/** Calls {@link event.listen}, with stronger typing for `id`. */
export function tauriListen(
  id:
    | 'reload'
    | 'new-note'
    | 'delete-note'
    | 'push-notes'
    | 'login'
    | 'logout'
    | 'signup'
    | `tauri://${string}`,
  cb: EventCallback<unknown>
): Promise<UnlistenFn> {
  return event.listen(id, cb);
}

/** Calls {@link invoke}, with stronger typing for `cmd`. */
export function tauriInvoke<T>(
  cmd: TauriCommand,
  args?: Parameters<typeof invoke>[1]
): Promise<T> {
  return invoke(cmd, args);
}
