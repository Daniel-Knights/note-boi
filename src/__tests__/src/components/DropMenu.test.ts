import { mount } from '@vue/test-utils';

import { DropMenuItemData } from '../../../components/types';
import { mockApi } from '../../mock';
import { findByTestId, getByTestId } from '../../utils';

import DropMenu from '../../../components/DropMenu.vue';

const ITEM_CLASS = 'drop-menu__item';

function mountDropMenu(items: DropMenuItemData[] = []) {
  return mount(DropMenu, {
    props: { items },
  });
}

describe('DropMenu', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();

    const wrapper = mountDropMenu();

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);
  });

  it('Emits close on click', () => {
    const wrapper = mountDropMenu();

    document.body.click();

    assert.lengthOf(wrapper.emitted('close')!, 1);
  });

  it('Renders each item', () => {
    const items = Array.from({ length: 5 }, (_, i) => ({
      label: `label-${i}`,
      testId: `item-${i}`,
    }));
    const wrapper = mountDropMenu(items);

    items.forEach(({ label, testId }) => {
      const itemWrapper = getByTestId(wrapper, testId);

      assert.isTrue(itemWrapper.isVisible());
      assert.strictEqual(itemWrapper.text(), label);
    });
  });

  it('Adds click handlers to items', () => {
    const items = Array.from({ length: 5 }, (_, i) => ({
      label: '',
      testId: `item-${i}`,
      clickHandler: vi.fn(),
    }));
    const wrapper = mountDropMenu(items);

    items.forEach(({ testId, clickHandler }) => {
      const itemWrapper = getByTestId(wrapper, testId);
      assert.isTrue(itemWrapper.isVisible());

      itemWrapper.trigger('click');

      expect(clickHandler).toHaveBeenCalledOnce();
    });
  });

  it('Adds classes to items', () => {
    const items = Array.from({ length: 5 }, (_, i) => ({
      label: '',
      testId: `item-${i}`,
      disabled: true,
      selected: true,
      danger: true,
      subMenu: [],
    }));
    const wrapper = mountDropMenu(items);

    items.forEach(({ testId }) => {
      const itemWrapper = getByTestId(wrapper, testId);

      assert.isTrue(itemWrapper.isVisible());
      assert.isTrue(itemWrapper.classes(`${ITEM_CLASS}--disabled`));
      assert.isTrue(itemWrapper.classes(`${ITEM_CLASS}--selected`));
      assert.isTrue(itemWrapper.classes(`${ITEM_CLASS}--danger`));
      assert.isTrue(itemWrapper.classes(`${ITEM_CLASS}--has-sub-menu`));
    });
  });

  it('Handles sub-menus', () => {
    const items = Array.from({ length: 5 }, (_, i) => ({
      label: '',
      testId: `item-${i}`,
      subMenu: [
        { label: '', testId: 'sub-item-1' },
        { label: '', testId: 'sub-item-2' },
        { label: '', testId: 'sub-item-3' },
      ],
    }));
    const wrapper = mountDropMenu(items);

    items.forEach(({ testId }) => {
      const itemWrapper = getByTestId(wrapper, testId);

      assert.isTrue(itemWrapper.isVisible());
      assert.isTrue(itemWrapper.classes(`${ITEM_CLASS}--has-sub-menu`));
      assert.isTrue(findByTestId(itemWrapper, 'sub-item-1').exists());
      assert.isTrue(findByTestId(itemWrapper, 'sub-item-2').exists());
      assert.isTrue(findByTestId(itemWrapper, 'sub-item-3').exists());
    });
  });
});
