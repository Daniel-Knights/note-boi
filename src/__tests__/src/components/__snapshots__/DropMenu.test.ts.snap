// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropMenu > Adds classes to items > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Adds classes to items > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Adds classes to items > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Adds classes to items > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Adds classes to items > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Adds classes to items > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Adds classes to items > allCalls 1`] = `[]`;

exports[`DropMenu > Adds classes to items > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Adds classes to items > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Adds classes to items > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Adds classes to items > openedPopup 1`] = `undefined`;

exports[`DropMenu > Adds classes to items > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Adds classes to items > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Adds classes to items > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DropMenu > Adds click handlers to items > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Adds click handlers to items > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Adds click handlers to items > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Adds click handlers to items > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Adds click handlers to items > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Adds click handlers to items > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Adds click handlers to items > allCalls 1`] = `[]`;

exports[`DropMenu > Adds click handlers to items > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Adds click handlers to items > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Adds click handlers to items > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Adds click handlers to items > openedPopup 1`] = `undefined`;

exports[`DropMenu > Adds click handlers to items > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Adds click handlers to items > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Adds click handlers to items > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DropMenu > Emits close on click > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Emits close on click > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Emits close on click > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Emits close on click > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Emits close on click > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Emits close on click > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Emits close on click > allCalls 1`] = `[]`;

exports[`DropMenu > Emits close on click > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Emits close on click > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Emits close on click > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Emits close on click > openedPopup 1`] = `undefined`;

exports[`DropMenu > Emits close on click > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Emits close on click > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Emits close on click > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DropMenu > Handles sub-menus > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Handles sub-menus > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Handles sub-menus > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Handles sub-menus > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Handles sub-menus > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Handles sub-menus > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Handles sub-menus > allCalls 1`] = `[]`;

exports[`DropMenu > Handles sub-menus > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Handles sub-menus > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Handles sub-menus > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Handles sub-menus > openedPopup 1`] = `undefined`;

exports[`DropMenu > Handles sub-menus > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Handles sub-menus > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Handles sub-menus > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DropMenu > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Mounts > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Mounts > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Mounts > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Mounts > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Mounts > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Mounts > allCalls 1`] = `[]`;

exports[`DropMenu > Mounts > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Mounts > openedPopup 1`] = `undefined`;

exports[`DropMenu > Mounts > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DropMenu > Renders each item > KeyStore.getKey() 1`] = `undefined`;

exports[`DropMenu > Renders each item > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DropMenu > Renders each item > Storage.get('THEME') 1`] = `null`;

exports[`DropMenu > Renders each item > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DropMenu > Renders each item > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DropMenu > Renders each item > Storage.get('USERNAME') 1`] = `null`;

exports[`DropMenu > Renders each item > allCalls 1`] = `[]`;

exports[`DropMenu > Renders each item > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DropMenu > Renders each item > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`DropMenu > Renders each item > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`DropMenu > Renders each item > openedPopup 1`] = `undefined`;

exports[`DropMenu > Renders each item > selectedTheme 1`] = `"System"`;

exports[`DropMenu > Renders each item > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`DropMenu > Renders each item > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
